---
title: "TSIV Genome Annotation"
author: "Alyssa Yoxsimer"
date: '2024-08-05'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# Load libraries
library(tidyverse); theme_set(theme_bw())
library(fs)
```
```{r}
# Input files
gff_file <- "../../analysis/assembly/03_annotate/geneMarkS_TSIV_STMY_X_2011_03_M_contigs/TSIV_STMY_X_2011_03_M_contigs_GeneMarkS.gff"
blastp_file <- "../../analysis/assembly/03_annotate/geneMarkS_TSIV_STMY_X_2011_03_M_contigs/TSIV_STMY_X_2011_03_M_contigs_GeneMarkS_blastp.tsv"
blastp_folder <- "/labs/kingsley/ambenj/TSIV/analysis/assembly/04_blastp_iridoviruses"
core_genes_TSIV_file <- "/labs/kingsley/ambenj/TSIV/analysis/assembly/04_blastp_iridoviruses/core_genes_TSIV.txt"

# Output files
annot_file <- "../../analysis/assembly/03_annotate/geneMarkS_TSIV_STMY_X_2011_03_M_contigs/TSIV_STMY_X_2011_03_M_contigs_GeneMarkS_annotations_needs_curation.tsv"
top_B22_gb_file <- "../../analysis/assembly/08_B22_phylogeny/top_B22_gb_accession_list.txt"
top_B22_gb_taxid_file <- "../../analysis/assembly/08_B22_phylogeny/top_B22_gb_taxid_accession_list.txt"
```

```{r}
# Load core gene file
core_genes <- read_tsv(core_genes_TSIV_file)
core_genes
```


```{r}
# Load gff file
gff <- read_tsv(gff_file, comment = "#", col_names = c(
        "seqid",
        "source",
        "type",
        "start",
        "stop",
        "score",
        "strand",
        "phase",
        "attr"
    )) %>% 
  separate(attr, sep = ',', into = c("gene_id"), extra = "drop", remove = FALSE) %>% 
  mutate(gene_id = gsub('gene_id=', "gene_", gene_id),
         orf = as.integer(str_remove(gene_id, "gene_")),
         NODE_temp = str_remove(seqid, "NODE_")) %>% 
  separate(NODE_temp, sep="_", into=c("NODE"), remove = TRUE) %>% 
  mutate(NODE = as.numeric(NODE))

gff
```
```{r}
# Get number of predicted ORFs on each sequence
gff %>% 
  group_by(NODE) %>% 
  summarize(n=n()) %>% 
  arrange(NODE)
```

# Top blastp hits from full database search
```{r}
# Read blastp results
blastp <- read_tsv(blastp_file, col_names = c("qseqid", "sseqid", "evalue", "stitle", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "bitscore", "sscinames", "scomnames", "staxid", "sacc")) %>% 
  separate(qseqid, sep = "\\|", into=c("gene_id"), extra = "drop", remove = FALSE) %>% 
  filter(evalue < 1e-02) %>% 
  left_join(.,gff, by = join_by(gene_id))
blastp
```
```{r}
# Get top result for genes with iridovirus hit
blastp_iridovirus_top <- blastp %>%  
  filter(grepl('iridovirus', stitle)) %>% 
  group_by(gene_id) %>% 
  arrange(evalue, -bitscore) %>% 
  slice_head(n=1) %>%
  ungroup() %>% 
  arrange(orf)
blastp_iridovirus_top
```
```{r}
# Find contigs with genes that have iridovirus hits
blastp_iridovirus_top %>% 
  group_by(seqid) %>% 
  summarise(n =n()) %>% 
  arrange(-n)
```
```{r}
# Counts number of results by gene_id
blastp %>%
  group_by(gene_id) %>% 
  summarise(n=n()) %>% 
  arrange(-n)
```

```{r}
# Append top blast hit to annotation table
annot <- read_tsv(blastp_file, col_names = c("qseqid", "sseqid", "evalue", "stitle", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "bitscore", "sscinames", "scomnames", "staxid", "sacc")) %>% 
  separate(qseqid, sep = "\\|", into=c("gene_id"), extra = "drop", remove = FALSE) %>% 
  group_by(gene_id) %>% 
  arrange(evalue, -bitscore) %>% 
  slice(1) %>% 
  ungroup() %>% 
  left_join(gff, .)

annot
```

# Annotations from selected iridovirus genomes
```{r}
# Read blastp results files
files <- dir_ls(path = blastp_folder, glob = "*_blastp.tsv")

# function to add file name to dataframe
read_and_record_filename <- function(filename){
  read_tsv(filename, col_names = c("qseqid", "sseqid", "evalue", "stitle", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "bitscore")) %>%
  mutate(filename = path_file(filename))
  }

# gather files into one dataframe
blastp_IV_genomes <- map_df(files, read_and_record_filename) %>% 
  separate(qseqid, into=c("TSIV_ORF"), sep="\\|", remove=FALSE, extra="drop") %>% 
  mutate(TSIV_ORF = as.integer(str_remove(TSIV_ORF, "gene_"))) %>% 
  separate(filename, into=c("virus_accession"), sep="\\.", extra="drop")
blastp_IV_genomes %>% 
  arrange(TSIV_ORF)
```

```{r}
# Filter to top hit from each iridovirus genome
top_blastp_IV_genomes <- blastp_IV_genomes %>% 
  filter(evalue < 1e-02) %>% 
  group_by(TSIV_ORF, virus_accession) %>% 
  arrange(evalue, -bitscore) %>% 
  slice(1) %>% 
  ungroup()
```

```{r}
# Get corresponding ISKNV ORF and clean up "ECIV-LEC15001_MK637631", "SDDV-C4575_KR139659"
ISKNV <- top_blastp_IV_genomes %>% 
  filter(virus_accession == "ISKNV_AF371960") %>% 
  separate(stitle, into=c("ISKNV_AF371960_ORF"), sep='\\(', remove=FALSE, extra="drop") %>% 
  select(TSIV_ORF, ISKNV_AF371960_ORF)
```

```{r}
# Get corresponding ISKNV ORF and clean up "ECIV-LEC15001_MK637631", "SDDV-C4575_KR139659"
SDDV <- top_blastp_IV_genomes %>% 
  filter(virus_accession == "SDDV-C4575_KR139659") %>% 
  separate(stitle, into=c("skip1", "skip2", "SDDV-C4575_KR139659"), sep='ORF_|]', remove=FALSE, extra="drop") %>% 
  mutate(`SDDV-C4575_KR139659` = str_replace(`SDDV-C4575_KR139659`, "^0+|^O+", "")) %>% 
  select(TSIV_ORF, `SDDV-C4575_KR139659`)
SDDV
```

```{r}
# Get corresponding ISKNV ORF and clean up "ECIV-LEC15001_MK637631", "SDDV-C4575_KR139659"
ECIV <- top_blastp_IV_genomes %>% 
  filter(virus_accession == "ECIV-LEC15001_MK637631") %>% 
  separate(stitle, into=c("skip1", "ECIV-LEC15001_MK637631"), sep='ORF|]', remove=FALSE, extra="drop") %>% 
  select(TSIV_ORF, `ECIV-LEC15001_MK637631`)
```

```{r}
# Add three comparisons to annotatation table
annot <- left_join(annot, ISKNV, by=c("orf"="TSIV_ORF")) %>% 
  left_join(., SDDV, by=c("orf"="TSIV_ORF")) %>% 
  left_join(., ECIV, by=c("orf"="TSIV_ORF"))
annot
```



```{r}
# Get filtered table
final_annot <- annot %>% 
  left_join(., filter(core_genes, !is.na(qseqid))) %>% 
  mutate(type = case_when(!is.na(`Gene Name`) ~ "Iridovirus core gene",
                          is.na(sseqid) ~ "Unknown",
                          !is.na(ISKNV_AF371960_ORF) | !is.na(`SDDV-C4575_KR139659`) | !is.na(`ECIV-LEC15001_MK637631`) | !str_detect(stitle, "iridovirus")  ~ "Iridovirus shared gene",
                          TRUE ~ NA)) %>% 
  arrange(orf) %>% 
  select(orf, seqid, start, stop, strand, type, `Gene Name`, stitle, sacc, evalue, ISKNV_AF371960_ORF, `SDDV-C4575_KR139659`, `ECIV-LEC15001_MK637631`)

final_annot
```
```{r}
# Write annotation outfile
final_annot %>% 
  write_tsv(annot_file)
```


# B22 analysis

```{r}
# Get annotations for top B22 (gene_90) Blastp hits from genbank
top_B22_gb <- blastp %>% 
  filter(gene_id=="gene_90", str_starts(sseqid, 'gb')) %>% 
  arrange(evalue, -bitscore) %>% 
  slice_head(n=75)
top_B22_gb

```

```{r}
# Get annotations for top B22 (gene_90) Blastp hits from genbank, and pick best hit from each taxid and write to text file
top_B22_gb_taxid <- blastp %>% 
  filter(gene_id=="gene_90", str_starts(sseqid, 'gb')) %>% 
  group_by(staxid) %>% 
  arrange(evalue, -bitscore) %>% 
  slice(1) %>% 
  ungroup() %>% 
  arrange(evalue, -bitscore)

top_B22_gb_taxid
```

```{r}
top_B22_gb %>% 
  select(sacc, stitle) %>% 
  write_tsv(top_B22_gb_file)

top_B22_gb_taxid %>% 
  select(sacc, stitle) %>% 
  write_tsv(top_B22_gb_taxid_file)
```




