---
title: "TSIV Genome Annotation"
author: "Alyssa Yoxsimer"
date: '2024-08-05'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# Load libraries
library(tidyverse); theme_set(theme_bw())
```
```{r}
# Input files
gff_file <- "../../analysis/assembly/03_annotate/geneMarkS_STMY_X_2011_03_M_all/STMY_X_2011_03_M_all_GeneMarkS.gff"
blastn_file <- "../../analysis/assembly/03_annotate/geneMarkS_STMY_X_2011_03_M_all/STMY_X_2011_03_M_all_GeneMarkS_blastn.tsv"
```

```{r}
# Load gff file
gff <- read_tsv(gff_file, comment = "#", col_names = c(
        "seqid",
        "source",
        "type",
        "start",
        "stop",
        "score",
        "strand",
        "phase",
        "attr"
    )) %>% 
  separate(attr, sep = ',', into = c("gene_id"), extra = "drop", remove = FALSE) %>% 
  mutate(gene_id = gsub('gene_id=', "gene_", gene_id))

gff
```

```{r}
# Read blastn results
blastn <- read_tsv(blastn_file, col_names = c("qseqid", "sseqid", "evalue", "stitle", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "bitscore", "sscinames", "scomnames", "staxid", "sacc")) %>% 
  separate(qseqid, sep = "\\|", into=c("gene_id"), extra = "drop", remove = FALSE) %>% 
  filter(evalue < 1e-02)
blastn
```
```{r}
# Get top result for genes with iridovirus hit
blastn_iridovirus_top <- blastn %>%  
  filter(grepl('iridovirus', stitle)) %>% 
  mutate(gene_num = as.numeric(str_remove(gene_id, "gene_"))) %>% 
  arrange(-gene_num) %>% 
  left_join(.,gff, by = join_by(gene_id)) %>% 
  group_by(gene_id) %>% 
  arrange(-bitscore) %>% 
  slice_head(n=1) %>%
  ungroup()
blastn_iridovirus_top
```
```{r}
# Find contigs with genes that have iridovirus hits
blastn_iridovirus_top %>% 
  group_by(seqid) %>% 
  summarise(n =n()) %>% 
  arrange(-n)
```


```{r}
# How many genes are annotated for NODE_8?
gff %>% 
    filter(grepl('NODE_8_', seqid))

# What blast hits does that gene have?
blastn %>% 
  filter(grepl('gene_99\\|', qseqid)) %>% 
  arrange(-bitscore)
```  

```{r}
# How many genes are annotated for NODE_18?
gff %>% 
    filter(grepl('NODE_18_', seqid))

# What blast hits does that gene_105 have?
blastn %>% 
  filter(grepl('gene_105\\|', qseqid)) %>% 
  arrange(-bitscore)

# What blast hits does that gene_105 have?
blastn %>% 
  filter(grepl('gene_106\\|', qseqid)) %>% 
  arrange(-bitscore)
```

NODE_1, NODE_2, and NODE_3 are definitely iridovirus genome sequence. They have a consistent level level of coverage (~15-16x depth when reads are remapped). 
 
NODE_8 is 2597 bp long and has only one ORF annotated (gene_99). The coverage is also much higher than the other iridovirus contigs (~97x). The nucleotide sequence has strong blast hits for a few fish genome assemblies (lantern fish, European smelt, large-eye snaggletooth) followed by hits to a European chub iridovirus isolate and scale drop disease virus isolate. From a BLASTP search, this is a B22 protein (same one that we have been interested in!). Could there be multiple copies of this B22 in the iridovirus genome???

NODE_18 is 2146 bp long and has two ORFs annotated (gene_105, gene_106). The read depth is consistent with the first three iridovirus contigs (~15.5x). The gene_105 nucleotide sequence has the strongest blast hits for iridovirus genomes but interestingly also has a hit for "PREDICTED: Gasterosteus aculeatus aculeatus ribonucleoside-diphosphate reductase subunit M2 (LOC120815250), mRNA". The strongest protein blast hits are for ribonucleotide reductase beta/small subunits from iridoviruses. This does seem like perhaps a host gene that was picked up in iridoviruses. The gene_106 nucleotide sequence has no blast hits but the protein sequence does have hits for lamin-type epidermal growth factors found in other iridoviruses. Overall it does seem like NODE_18 should be considered part of the iridovirus reference genome.






